console/collections/src/merkle_tree/mod.rs

```diff
@@ -208,6 +208,206 @@ impl<E: Environment, LH: LeafHash<Hash = PH::Hash>, PH: PathHash<Hash = Field<E>
+        let (first, rest) = updates.split_first().unwrap();
+
+        // Assign to the most recently seen leaf index.
+        let mut latest_leaf_index = first.0;
+
+        // Check that the latest leaf index is less than number of leaves in the Merkle tree.
+        ensure!(
+            latest_leaf_index < self.number_of_leaves,
+            "Leaf index must be less than the number of leaves in the Merkle tree"
+        );
+
+        // Check that the rest of the leaf hashes are in descending order.
+        for (leaf_index, _) in rest {
+            ensure!(*leaf_index < latest_leaf_index, "Leaf indices must be sorted in strictly descending order");
+            // Update the latest leaf index.
+            latest_leaf_index = *leaf_index;
+        }
+
+        // Compute the start index (on the left) for the leaf hashes level in the Merkle tree.
+        let start = match self.number_of_leaves.checked_next_power_of_two() {
+            Some(num_leaves) => num_leaves - 1,
+            None => bail!("Integer overflow when computing the Merkle tree start index"),
+        };
+
+        // A helper to compute the leaf hash.
+        let hash_update = |(leaf_index, leaf): &(usize, LH::Leaf)| {
+            self.leaf_hasher.hash_leaf(leaf).map(|hash| (start + leaf_index, hash))
+        };
+
+        // Hash the leaves and add them to the updated hashes.
+        let leaf_hashes: Vec<(usize, LH::Hash)> = match updates.len() {
+            0..=100 => updates.iter().map(|update| hash_update(update)).collect::<Result<Vec<_>>>()?,
+            _ => cfg_iter!(updates).map(|update| hash_update(update)).collect::<Result<Vec<_>>>()?,
+        };
+        lap!(timer, "Hashed {} new leaves", leaf_hashes.len());
+
+        // Store the updated hashes by level.
+        let mut updated_hashes = vec![leaf_hashes];
+
+        // A helper function to compute the path hashes for a given level.
+        type Update<PH> = (usize, (<PH as PathHash>::Hash, <PH as PathHash>::Hash));
+        let compute_path_hashes = |inputs: &[Update<PH>]| match inputs.len() {
+            0..=100 => inputs
+                .iter()
+                .map(|(index, (left, right))| self.path_hasher.hash_children(left, right).map(|hash| (*index, hash)))
+                .collect::<Result<Vec<_>>>(),
+            _ => cfg_iter!(inputs)
+                .map(|(index, (left, right))| self.path_hasher.hash_children(left, right).map(|hash| (*index, hash)))
+                .collect::<Result<Vec<_>>>(),
+        };
+
+        // Compute the depth of the tree. This corresponds to the number of levels of hashes in the tree.
+        let tree_depth = tree_depth::<DEPTH>(self.tree.len())?;
+        // For each level in the tree, compute the path hashes.
+        for level in 0..tree_depth as usize {
+            // Prepare the inputs to path hasher.
+            let mut inputs = Vec::with_capacity(updated_hashes[level].len());
```

Comment:
this vector could be allocated once before the loop, `clear`ed after iteration, and reused